pkg/netutil/cni_plugin.go

```diff
@@ -0,0 +1,171 @@
+}
+
+func (*natConfig) GetPluginType() string {
+	return "nat"
+}
+
+func newNatPlugin(master string) *natConfig {
+	return &natConfig{
+		PluginType: "nat",
+		Master:     master,
+	}
+}
+
+// portMapConfig describes the portmapping plugin
+type portMapConfig struct {
+	PluginType   string          `json:"type"`
+	Capabilities map[string]bool `json:"capabilities"`
+}
+
+func newPortMapPlugin() *portMapConfig {
+	return &portMapConfig{
+		PluginType: "portmap",
+		Capabilities: map[string]bool{
+			"portMappings": true,
+		},
+	}
+}
+
+func (*portMapConfig) GetPluginType() string {
+	return "portmap"
+}
+
+// firewallConfig describes the firewall plugin
+type firewallConfig struct {
+	PluginType string `json:"type"`
+	Backend    string `json:"backend,omitempty"`
+}
+
+func newFirewallPlugin() *firewallConfig {
+	return &firewallConfig{
+		PluginType: "firewall",
+	}
+}
+
+func (*firewallConfig) GetPluginType() string {
+	return "firewall"
+}
+
+// tuningConfig describes the tuning plugin
+type tuningConfig struct {
+	PluginType string `json:"type"`
+}
+
+func newTuningPlugin() *tuningConfig {
+	return &tuningConfig{
+		PluginType: "tuning",
+	}
+}
+
+func (*tuningConfig) GetPluginType() string {
+	return "tuning"
+}
+
+type isolationConfig struct {
+	PluginType string `json:"type"`
+}
+
+func newIsolationPlugin() *isolationConfig {
+	return &isolationConfig{
+		PluginType: "isolation",
+	}
+}
+
+func (*isolationConfig) GetPluginType() string {
+	return "isolation"
+}
+
+// https://github.com/containernetworking/plugins/blob/v1.0.1/plugins/ipam/host-local/backend/allocator/config.go#L47-L56
+type hostLocalIPAMConfig struct {
+	Type        string      `json:"type"`
+	Routes      []IPAMRoute `json:"routes,omitempty"`
+	ResolveConf string      `json:"resolveConf,omitempty"`
+	DataDir     string      `json:"dataDir,omitempty"`
+	Ranges      [][]IPRange `json:"ranges,omitempty"`
+}
+
+type IPRange struct {
+	Subnet     string `json:"subnet"`
+	RangeStart string `json:"rangeStart,omitempty"`
+	RangeEnd   string `json:"rangeEnd,omitempty"`
+	Gateway    string `json:"gateway,omitempty"`
+}
+
+type IPAMRoute struct {
+	Dst     string `json:"dst,omitempty"`
+	GW      string `json:"gw,omitempty"`
+	Gateway string `json:"gateway,omitempty"`
+}
+
+func newHostLocalIPAMConfig() *hostLocalIPAMConfig {
+	return &hostLocalIPAMConfig{
+		Type: "host-local",
+	}
+}
+
+// https://github.com/microsoft/windows-container-networking/blob/v0.2.0/cni/cni.go#L55-L63
+type windowsIpamConfig struct {
+	Type          string      `json:"type"`
+	Environment   string      `json:"environment,omitempty"`
+	AddrSpace     string      `json:"addressSpace,omitempty"`
+	Subnet        string      `json:"subnet,omitempty"`
+	Address       string      `json:"ipAddress,omitempty"`
+	QueryInterval string      `json:"queryInterval,omitempty"`
+	Routes        []IPAMRoute `json:"routes,omitempty"`
+}
+
+func newWindowsIPAMConfig() *windowsIpamConfig {
+	return &windowsIpamConfig{}
+}
```

Comment:
Can we move windows-specific stuffs to `*_windows.go`, and linux-specific stuffs to `*_linux.go`?

Reply:
Here is just some structural definitions, separation will make tricky for `GenerateCNIPlugins`.